File: src/com/timepath/hl2/cvars/CVarList.java
package com.timepath.hl2.cvars;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Map;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * http://gskinner.com/RegExr/
 *
 * @author TimePath
 */
public class CVarList {

    private static final Logger LOG = Logger.getLogger(CVarList.class.getName());

    public static Map<String, CVar> analyzeList(Scanner scanner, Map<String, CVar> map) {
        CVar c = null;
        Pattern cvarlist = Pattern.compile(
                "([\\S]*)[\\s]*:[\\s]*([\\S]*)[\\s]*:[^\\\"]*(.*)[\\s]*:[\\s]*(.*)");
        Pattern tag = Pattern.compile("\\\"([^\\\"]*)\\\"");
        Pattern kv = Pattern.compile("\\\"([^\"]*)\\\"[\\s]*=[\\s]*\\\"([^\"]*)\\\"[\\s]*(.*)");
        Pattern desc = Pattern.compile("^(?: - | )(.*)");
        Pattern defaultValue = Pattern.compile("\\([^\"]*\"(.*)\"[^\"]*\\)");
        Pattern minValue = Pattern.compile("(?:min\\.\\s)(.*?)(?:\\s|$)");
        Pattern maxValue = Pattern.compile("(?:max\\.\\s)(.*?)(?:\\s|$)");
        while(scanner.hasNext()) {
            String line = scanner.nextLine();
            if(line.trim().length() == 0) {
                continue;
            }
            LOG.fine(line);
            Matcher cvarlistMatcher = cvarlist.matcher(line);
            Matcher kvMatcher = kv.matcher(line);
            Matcher descMatcher = desc.matcher(line);
            if(kvMatcher.find()) {
//                    LOG.info("KV match");
                String name = kvMatcher.group(1);
                Object value = kvMatcher.group(2);
                if(!map.containsKey(name)) {
                    c = new CVar();
                    c.setName(name);
                    c.setValue(value);
                } else {
                    c = map.get(name);
                }
                String extra = kvMatcher.group(3);
                if(extra.length() > 0) {
                    Matcher defaultMatcher = defaultValue.matcher(extra);
                    if(defaultMatcher.find()) {
                        c.setDefaultValue(defaultMatcher.group(1));
                    }
                    Matcher minMatcher = minValue.matcher(extra);
                    if(minMatcher.find()) {
                        c.setMinimum(minMatcher.group(1));
                    }
                    Matcher maxMatcher = maxValue.matcher(extra);
                    if(maxMatcher.find()) {
                        c.setMaximum(maxMatcher.group(1));
                    }
                }
            } else if(cvarlistMatcher.find()) {
//                    LOG.info("cvarlist match");
                c = new CVar();
                c.setName(cvarlistMatcher.group(1));
                c.setValue(cvarlistMatcher.group(2));
                Matcher tagMatcher = tag.matcher(cvarlistMatcher.group(3));
                while(tagMatcher.find()) {
                    c.getTags().add(tagMatcher.group());
                }
                c.setDesc(cvarlistMatcher.group(4));
            } else if(descMatcher.find()) {
//                    LOG.info("Additional info match");
                if(c == null) {
                    LOG.warning("Data before a cvar");
                    continue;
                }
                String trimmed = descMatcher.group(1);
                boolean description = line.startsWith(" - ");
                if(description) {
                    c.setDesc(trimmed);
                } else {
                    c.getTags().addAll(Arrays.asList(trimmed.split(" ")));
                }
            } else {
                LOG.log(Level.INFO, "Unparsed: {0}", line);
                continue;
            }
            if(map.containsKey(c.getName())) {
                LOG.log(Level.INFO, "Duplicate: {0}", line);
            }
            map.put(c.getName(), c);
        }
        return map;
    }

    private CVarList() {
    }

    public static class CVar {

        private String name;

        /**
         * null if cmd
         */
        private Object value;

        private Object defaultValue;

        private Object minimum;

        private Object maximum;

        private String desc;

        private final ArrayList<String> tags = new ArrayList<String>();

        public void setName(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public void setValue(Object value) {
            this.value = value;
        }

        public Object getValue() {
            return value;
        }

        public void setDefaultValue(Object defaultValue) {
            this.defaultValue = defaultValue;
        }

        public Object getDefaultValue() {
            return defaultValue;
        }

        public Object getMinimum() {
            return minimum;
        }

        public void setMinimum(Object minimum) {
            this.minimum = minimum;
        }

        public Object getMaximum() {
            return maximum;
        }

        public void setMaximum(Object maximum) {
            this.maximum = maximum;
        }

        public void setDesc(String desc) {
            this.desc = desc;
        }

        public String getDesc() {
            return desc;
        }

        public ArrayList<String> getTags() {
            return tags;
        }

        @Override
        public boolean equals(Object obj) {
            return toString().equals(obj.toString());
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            sb.append(name).append(" : ").append(value).append(" : ");
            for(String tag : tags) {
                sb.append(", ").append("\"").append(tag).append("\"");
            }
            sb.append(" : ").append(desc);
            return sb.toString();
        }

    }

}

File: src/main/kotlin/com/timepath/hl2/cvar/CVarList.kt
package com.timepath.hl2.cvar

import java.util.Arrays
import java.util.Scanner
import java.util.logging.Level
import java.util.logging.Logger
import java.util.regex.Pattern

/**
 * @author TimePath
 */
object CVarList {

    private val LOG = Logger.getLogger(javaClass<CVarList>().getName())

    public fun analyzeList(scanner: Scanner, map: MutableMap<String, CVar>): Map<String, CVar> {
        var c: CVar? = null
        val cvarlist = Pattern.compile("([\\S]*)[\\s]*:[\\s]*([\\S]*)[\\s]*:[^\"]*(.*)[\\s]*:[\\s]*(.*)")
        val tag = Pattern.compile("\"([^\"]*)\"")
        val kv = Pattern.compile("\"([^\"]*)\"[\\s]*=[\\s]*\"([^\"]*)\"[\\s]*(.*)")
        val desc = Pattern.compile("^(?: - | )(.*)")
        val defaultValue = Pattern.compile("\\([^\"]*\"(.*)\"[^\"]*\\)")
        val minValue = Pattern.compile("(?:min\\.\\s)(.*?)(?:\\s|$)")
        val maxValue = Pattern.compile("(?:max\\.\\s)(.*?)(?:\\s|$)")
        while (scanner.hasNext()) {
            val line = scanner.nextLine()
            if (line.trim().isEmpty()) {
                continue
            }
            LOG.fine(line)
            val cvarlistMatcher = cvarlist.matcher(line)
            val kvMatcher = kv.matcher(line)
            val descMatcher = desc.matcher(line)
            if (kvMatcher.find()) {
                //                    LOG.info("KV match");
                val name = kvMatcher.group(1)
                val value = kvMatcher.group(2)
                if (!map.containsKey(name)) {
                    c = CVar()
                    c.name = name
                    c.value = value
                } else {
                    c = map[name]
                }
                val extra = kvMatcher.group(3)
                if (!extra.isEmpty()) {
                    val defaultMatcher = defaultValue.matcher(extra)
                    if (defaultMatcher.find()) {
                        c!!.defaultValue = defaultMatcher.group(1)
                    }
                    val minMatcher = minValue.matcher(extra)
                    if (minMatcher.find()) {
                        c!!.minimum = minMatcher.group(1)
                    }
                    val maxMatcher = maxValue.matcher(extra)
                    if (maxMatcher.find()) {
                        c!!.maximum = maxMatcher.group(1)
                    }
                }
            } else if (cvarlistMatcher.find()) {
                //                    LOG.info("cvarlist match");
                c = CVar()
                c.name = cvarlistMatcher.group(1)
                c.value = cvarlistMatcher.group(2)
                val tagMatcher = tag.matcher(cvarlistMatcher.group(3))
                while (tagMatcher.find()) {
                    c.tags.add(tagMatcher.group())
                }
                c.desc = cvarlistMatcher.group(4)
            } else if (descMatcher.find()) {
                //                    LOG.info("Additional info match");
                if (c == null) {
                    LOG.warning("Data before a cvar")
                    continue
                }
                val trimmed = descMatcher.group(1)
                val description = line.startsWith(" - ")
                if (description) {
                    c.desc = trimmed
                } else {
                    c.tags.addAll(trimmed.splitBy(" "))
                }
            } else {
                LOG.log(Level.INFO, "Unparsed: {0}", line)
                continue
            }
            if (map.containsKey(c!!.name)) {
                LOG.log(Level.INFO, "Duplicate: {0}", line)
            }
            map.put(c.name!!, c)
        }
        return map
    }
}
